### PR TITLE
Add Bloom filter network messages

### PR DESCRIPTION
**Pull request by benthecarman**
Adds initial support for parsing the `merkleblock`, `filterload`, `filteradd` and `filterclear` network messages.

https://developer.bitcoin.org/reference/p2p_networking.html#filteradd

https://developer.bitcoin.org/reference/p2p_networking.html#filterclear

### CODE REVIEW

**Comment by jrawsthorne** [comment]
Looks like other implementations represent the element in filteradd as raw bytes. The example you linked shows a txid being sent as an element which isn't a script. I worked on this a while ago but never created a pr. https://github.com/jrawsthorne/rust-bitcoin/blob/master/src/network/message_bloom.rs contains filterload as well if you want to include that as well

**Comment by benthecarman** [comment]
> Looks like other implementations represent the element in filteradd as raw bytes. The example you linked shows a txid being sent as an element which isn't a script. I worked on this a while ago but never created a pr. https://github.com/jrawsthorne/rust-bitcoin/blob/master/src/network/message_bloom.rs contains filterload as well if you want to include that as well

You're right, thanks

**Comment by benthecarman** [comment]
Implemented @jrawsthorne's suggestions, added him as a co-author too :)

**Review by jrawsthorne — DISMISSED** [review]
Followed https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki to double check everything is correct.

File: src/network/message.rs [hunk 1]
```diff
@@ -147,7 +148,14 @@ pub enum NetworkMessage {
     Ping(u64),
     /// `pong`
     Pong(u64),
-    // TODO: bloom filtering
+    /// `merkleblock`
+    MerkleBlock(MerkleBlock),
```
Comment (sgeisler) [inline src/network/message.rs:152]: Maybe add the bip number to the docstring like it was done for the BIP157 messages below?

**Review by sgeisler — APPROVED** [review]
utACK 0cf61adfd48c197ed1f6ede2986bd714dcfb258b

I'm operating under the assumption that the `MerkleBlock` implementation is correct, it looked ok at least. If that's the case this PR looks good.

File: src/network/message.rs (hunk 1 above)
Comment (benthecarman) [reply src/network/message.rs:152]: Done

**Review by sgeisler — DISMISSED** [review]
utACK c2135c976e0dbe798e357fd4264dea4f941a3b84

**Review by apoelstra — DISMISSED** [review]
ack c2135c976e0dbe798e357fd4264dea4f941a3b84

Reviewed the code and ran tests. Did not verify that the new structures match the spec.

Note that this is a breaking change.

**Review by apoelstra — DISMISSED** [review]
ACK e8904df5f36139ed9cea2725164a16b6fa0a96ef

Don't know why CI is not running. It passes locally (although I did not run the examples/ directory, and I have only access to x86_64, and not wasm)

**Comment by benthecarman** [comment]
> Don't know why CI is not running. It passes locally (although I did not run the examples/ directory, and I have only access to x86_64, and not wasm)

Github is weird with first time contributors and sometimes doesn't run CI

**Review by dr-orlovsky — DISMISSED** [review]
utACK e8904df5f36139ed9cea2725164a16b6fa0a96ef

Compared correspondence of data structures and serialization to the spec https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki#New_messages

**Comment by dr-orlovsky** [comment]
@benthecarman I tried everything to get the tests running so I can merge this PR. No success :( Seems like the only way is to force-push, which will discard all reviews however :( I will ask one of maintainers to add his confirmation ASAP if you will chose this path (I actually see no other route...)

**Comment by benthecarman** [comment]
> @benthecarman I tried everything to get the tests running so I can merge this PR. No success :( Seems like the only way is to force-push, which will discard all reviews however :( I will ask one of maintainers to add his confirmation ASAP if you will chose this path (I actually see no other route...)

Rebased

**Review by dr-orlovsky — APPROVED** [review]
ACK 894f0f09b609f73833affe4d09cac95cb51a0700

**Review by apoelstra — APPROVED** [review]
ACK 894f0f09b609f73833affe4d09cac95cb51a0700

**Pull request merged**